### ISSUE TITLE
Uri-compatible definitions of maps

### ISSUE THREAD

**Issue by ALEUT** [victools/jsonschema-generator#19]
I tried to test a schema created by this generator on site https://jsonschemalint.com/#!/version/draft-07/markup/json
It says `"#/definitions/Map<String, Set<String>>"` should match format "uri-reference". Looks it's because of space in map definition.

**Comment by CarstenWickner** [comment]
Hi @ALEUT,

Thanks for reporting this. A couple questions:
1. Does it work if you remove the white-space?
2. You don't happen to have a reference to the part of the json-schema specification where these limitations are described?

If simply removing all white-spaces is doing the trick, that'd be easy to fix.

**Comment by CarstenWickner** [comment]
Hi @ALEUT,

While I couldn't find an explicit description of valid/invalid characters in the `uri-reference` format I managed to reproduce the mentioned warning on jsonschemalint.com, which made me realize that this actually affects two kinds of references:
1. In the case of generic types with more than one type parameter, e.g. Map<String, String> (as per your example).
2. When specifically handling null-able definitions, which get the " (nullable)" suffix.

The simple fix here: removing any white-spaces in the definition keys, consequently not using those invalid keys as `$ref`.

The fix was now released as version [4.0.1](https://github.com/victools/jsonschema-generator/releases/tag/v4.0.1).

Thanks again for reporting. 👍

**Comment by ALEUT** [comment]
Hi @CarstenWickner 
1) Yes, it works if I remove white-space, but probably we need to check more deeply uri-reference documentation for format requirements.
2) Here https://json-schema.org/specification.html#single-vocabulary-meta-schemas you can download Core Vocabulary meta-schema (https://json-schema.org/draft/2019-09/meta/core) and it has such a definition:
```
"$ref": {
            "type": "string",
            "format": "uri-reference"
        }, 
```

And in https://json-schema.org/understanding-json-schema/reference/string.html I see
"uri-reference": New in draft 6 A URI Reference (either a URI or a relative-reference), according to RFC3986, section 4.1. https://tools.ietf.org/html/rfc3986#section-4.1

**Comment by ALEUT** [comment]
@CarstenWickner thanks for fix! When is it expected on mvnrepository.com?

**Comment by CarstenWickner** [comment]
Hi @ALEUT,

I released it two hours ago just before adding the above comment.
Might take a few hours to be propagated everywhere.
If you need it right away, you can download jar file from the release page here in GitHub. I usually upload it there as well.

I also found the same reference to the URI-Reference specification, but that didn’t help me much in understanding what are deemed invalid characters.
Since `<`, `>` and `,` seem to be okay, there shouldn’t be much else in type names that would cause this issue besides the white-spaces.
Maybe the `$` symbol for anonymous classes, but I’m not sure that can ever be picked-up via reflections.

I think we’re good for now.

**Comment by ALEUT** [comment]
Hi @CarstenWickner 

I also think it's good now. Thanks for the quick fix!

**Comment by ALEUT** [comment]
Hi @CarstenWickner 

New info regarding this bug:
I tried to use this library: https://github.com/everit-org/json-schema
And when tried to parse the schema built by your library, I got:
Exception in thread "main" java.lang.RuntimeException: java.net.URISyntaxException: Illegal character in fragment at index 17: #/definitions/Map<String,Set<TimeSeriesFeature>>

**Comment by CarstenWickner** [comment]
Hi @ALEUT,
Thanks for that. While I couldn’t find anything specific in the context of a JSON Schema, but for “URI” definitions in general there is this: https://tools.ietf.org/html/rfc3986#page-24

Apparently, valid characters are only:
- unreserved characters: a-z / A-Z / 0-9 / "-" / "." / "_" / "~"
- sub-delimiters: "!" / "$" / "&" / "'" / "(" / ")"
                  / "*" / "+" / "," / ";" / "="
- additionally allowed in fragments: ":" / “@“ / “/“ / “?”

Everything else needs to be percentage-encoded.
I.e. the angled brackets “<“ and “>” are not allowed. Probably best to replace them with characters that don’t require encoding, e.g. "(" and ")".

I’ll have a look later today.

**Comment by ALEUT** [comment]
@CarstenWickner what about [] for arrays?

**Comment by CarstenWickner** [comment]
Good catch. Maybe replacing the “[]” with a single “*” would convey that there are multiple.

Fortunately we are only talking about the technical references inside the schema and not necessarily about any title or description potentially being shown anywhere.

**Comment by CarstenWickner** [comment]
Hi @ALEUT,

Thanks again for reporting and following-up on this.
I've added your proposed check of `new URI(key)` to the tests generating example schemas, in addition to replacing `<`/`>`/`[]`.

These fixes are now released as 4.0.2.

**Comment by ALEUT** [comment]
@CarstenWickner verified using that library - works fine now!
